infrastructure/src/main/java/org/corfudb/infrastructure/AutoCommitService.java

```diff
@@ -0,0 +1,217 @@
+                new ThreadFactoryBuilder()
+                        .setDaemon(true)
+                        .setNameFormat(serverContext.getThreadPrefix() + "AutoCommitService")
+                        .build());
+    }
+
+    CorfuRuntime getCorfuRuntime() {
+        return runtimeSingletonResource.get();
+    }
+
+    /**
+     * Starts the long running service.
+     *
+     * @param interval interval to run the service
+     */
+    @Override
+    public void start(Duration interval) {
+        autoCommitScheduler.scheduleAtFixedRate(
+                () -> LambdaUtils.runSansThrow(this::runAutoCommit),
+                interval.toMillis() / 2,
+                interval.toMillis(),
+                TimeUnit.MILLISECONDS
+        );
+    }
+
+    @VisibleForTesting
+    public synchronized void runAutoCommit() {
+        boolean trimmed = false;
+        long lastEpoch = -1;
+        Layout currentLayout = null;
+
+        for (int i = 1; i <= MAX_COMMIT_RETRY; i++) {
+            try {
+                // Do not perform auto commit if current node is not primary sequencer.
+                currentLayout = updateLayoutAndGet();
+                if (!isNodePrimarySequencer(currentLayout)) {
+                    resetTails();
+                    return;
+                }
+
+                log.trace("runAutoCommit: start committing addresses.");
+
+                // Initialize lastLogTail and committedTail if the first time. We only
+                // commit up to the global tail at the time of last auto commit cycle.
+                if (!Address.isAddress(lastLogTail)) {
+                    initializeTails(currentLayout);
+                    return;
+                }
+
+                // Fetch maximum trim mark from log units and compare with committed tail.
+                // In order not to fetch trim mark for every retry, we only do this when
+                // the epoch changes after last attempt.
+                long currEpoch = currentLayout.getEpoch();
+                if (lastEpoch < 0 || trimmed || currEpoch != lastEpoch) {
+                    Token trimMark = getCorfuRuntime().getAddressSpaceView().getTrimMark(false);
+                    // Make sure all log units have same trim mark, then we can start committing
+                    // from the trim mark instead of the trailing committed tail.
+                    if (committedTail < trimMark.getSequence() - 1) {
+                        Token trimToken = new Token(trimMark.getEpoch(), trimMark.getSequence() - 1);
+                        getCorfuRuntime().getAddressSpaceView().prefixTrim(trimToken, false);
+                        committedTail = trimMark.getSequence() - 1;
+                    }
+                    lastEpoch = currEpoch;
+                    trimmed = false;
+                }
+
+                log.debug("runAutoCommit: trying to commit [{}, {}].", committedTail + 1, lastLogTail);
```

Comment:
please write a message to the log only ones - whether it commited or failed

Reply:
Done.